### PR TITLE
Add text-backbone support for Gemma 4 12B and 12B-it

### PR DESCRIPTION
**Pull request by deep-diver**
## Summary

This PR adds Tunix text-backbone support for:

- `google/gemma-4-12b`
- `google/gemma-4-12b-it`

It adds Gemma 4 12B model configs, registry entries, and safetensors mapping tests based on the existing Gemma 4 implementation.

This change covers the Gemma 4 12B language-model/text backbone only. Full Gemma 4 Unified multimodal image/audio/video support is out of scope.

## What Changed

- Added `ModelConfig.gemma4_12b()`
- Added `ModelConfig.gemma4_12b_it()`
- Registered:
  - `google/gemma-4-12b`
  - `google/gemma-4-12b-it`
- Added tests for:
  - config values
  - registry lookup
  - safetensors local/global attention mapping
  - AutoModel resolution

## Implementation Notes

The 12B config follows the HF `google/gemma-4-12B-it` text config:

- 48 layers
- hidden size 3840
- intermediate size 15360
- 16 attention heads
- 8 KV heads
- 1 global KV head
- global head dim 512
- sliding window 1024
- 5 local sliding layers followed by 1 global layer
- global attention uses `k_eq_v`

## Validation

Local tests:

```bash
pytest tests/models/gemma4/model_test.py tests/models/gemma4/params_safetensors_test.py
```

Result:

```text
10 passed
```

Registry / naming / AutoModel tests:

```text
586 passed, 61 deselected
```

GPU validation on A100-80GB:

- Loaded full `google/gemma-4-12B-it` safetensors checkpoint in Tunix/JAX
- Ran full forward pass
- Ran text generation through `tunix.generate.sampler`
- Verified HF-vs-Tunix short-prompt float32 parity
- Ran MMLU-Pro-style HF-vs-Tunix parity probes

Selected parity evidence:

```text
Short prompt float32 last-token logits cosine: 0.9999985
Top-50 overlap: 50/50
```

500-sample MMLU-Pro-style parity probe:

```text
Prediction match: 481/500 = 96.2%
Wilson 95% CI: 94.14% - 97.55%
HF accuracy: 75/500 = 15.0%
Tunix accuracy: 75/500 = 15.0%
Delta: 0.0 pp
```

Note: this is a benchmark-style parity probe, not an official MMLU-Pro benchmark reproduction.

Generation smoke tests:

- Arithmetic
- Explanation
- Code generation
- Structured extraction
- Gemma 4 thinking-template prompt

Tunix generated coherent text outputs and matched HF parsed outputs on representative text-only samples.

## Scope

This PR does not add full Gemma 4 Unified multimodal support.

The current Tunix Gemma4 model path supports text tokens only. The unified checkpoint contains multimodal weights such as:

```text
model.embed_audio.embedding_projection.weight
model.embed_vision.embedding_projection.weight
model.vision_embedder.*
```

These are intentionally out of scope for this PR and are not loaded by the current text-backbone implementation.

### CODE REVIEW

**Comment by tianshub** [comment]
@deep-diver thanks a lot for adding the support. A followup question: how is the MMLU result being measured? is it using Tunix's sampler? Meanwhile, can you also include a sampling output in the test plan?

**Comment by deep-diver** [comment]
Thanks for the comment @tianshub. To clarify, the MMLU result I mentioned was not
measured through `tunix.generate.sampler`. It was from a lightweight
HF-vs-Tunix parity/eval probe that I used to sanity-check the Gemma4 12B text
path. I should treat that as supporting validation evidence, not as an official
Tunix sampler benchmark.

Also, here is the some of the sample outputs obtained with `tunix.generate.sampler`.

Environment:

- A100 80GB
- JAX device: `cuda:0`
- dtype: `bfloat16`
- decoding: greedy, `temperature=0.0`, `top_p=None`
- implementation: this PR's Tunix Gemma4 model path

Representative actual outputs:

1. Arithmetic, `enable_thinking=False`

Prompt:

````text
Answer with only the final number. What is 17 plus 25, then minus 9?
````

Output:

````text
33
````

2. Explanation, `enable_thinking=False`

Prompt:

````text
Explain how a refrigerator cools the inside in exactly 3 sentences for a middle-school student.
````

Output:

````text
A refrigerator works by using a special fluid that evaporates into a gas as it absorbs heat from the food inside. This gas travels through coils and is compressed by a pump, which turns it back into a hot liquid. Finally, the liquid releases that heat into the room outside and cycles back inside to start the cooling process again.
````

3. Structured extraction, `enable_thinking=False`

Prompt:

````text
Return JSON only. Extract task, priority, and owner from: 'Alex, please organize the Gemma4 parity evidence by tomorrow. Priority is high.'
````

Output:

````text
```json
{
  "task": "organize the Gemma4 parity evidence",
  "priority": "high",
  "owner": "Alex"
}
```
````

4. Code generation, `enable_thinking=False`

Prompt:

````text
Write a Python function named top_k_words(text, k) that returns the k most frequent lowercase words. Keep it short.
````

Output:

````text
Here is a concise solution using the `collections.Counter` module and regular expressions to handle punctuation.

```python
import re
from collections import Counter

def top_k_words(text, k):
    words = re.findall(r'\w+', text.lower())
    return Counter(words).most_common(k)

# Example usage:
# print(top_k_words("Hello world! Hello again, world."), 2)
# Output: [('hello', 2), ('world', 2)]
```

### Explanation:
1.  **`re.findall(r'\w+', text.lower())`**: Converts the string to lowercase and extracts all alphanumeric sequences (words), automatically stripping punctuation.
2.  **`Counter(words)`**: Creates a dictionary-like object that counts the occurrences of each word.
3.  **`.most_common(k)`**: Returns a list of the $k$ most frequent elements as tuples.
````

5. Structured extraction, `enable_thinking=True`

Prompt:

````text
Return JSON only. Extract task, priority, and owner from: 'Alex, please organize the Gemma4 parity evidence by tomorrow. Priority is high.'
````

Output:

````text
<|channel>thought
*   Input: "Alex, please organize the Gemma4 parity evidence by tomorrow. Priority is high."
    *   Task: Extract "task", "priority", and "owner".
    *   Constraint: Return JSON only.

    *   Owner: Alex
    *   Task: organize the Gemma4 parity evidence
    *   Priority: high

    *   `task`: "organize the Gemma4 parity evidence"
    *   `priority`: "high"
    *   `owner`: "Alex"

    ```json
    {
      "task": "organize the Gemma4 parity evidence",
      "priority": "high",
      "owner": "Alex"
    }
    ```

    *   JSON only? Yes.
    *   Extracted correctly? Yes.<channel|>```json
{
  "task": "organize the Gemma4 parity evidence",
  "priority": "high",
  "owner": "Alex"
}
```
````

**Comment by tianshub** [comment]
Great! Thank you @deep-diver , please rebase and let's merge this.

**Review by tianshub — APPROVED** [review]

**Comment by deep-diver** [comment]
@tianshub thanks! I think it is ready to be merged now :)

**Comment by deep-diver** [comment]
@tianshub ping in case you missed my message
